src/rpc_client/eth_rpc/mod.rs

```diff
@@ -0,0 +1,851 @@
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "0x{}", hex::encode(self.0))
+    }
+}
+
+impl UpperHex for Hash {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "0x{}", hex::encode_upper(self.0))
+    }
+}
+
+impl std::str::FromStr for Hash {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if !s.starts_with("0x") {
+            return Err("Ethereum hash doesn't start with 0x".to_string());
+        }
+        let mut bytes = [0u8; 32];
+        hex::decode_to_slice(&s[2..], &mut bytes)
+            .map_err(|e| format!("failed to decode hash from hex: {}", e))?;
+        Ok(Self(bytes))
+    }
+}
+
+impl HttpResponsePayload for Hash {}
+
+/// Block tags.
+/// See <https://ethereum.org/en/developers/docs/apis/json-rpc/#default-block>
+#[derive(Debug, Default, Copy, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum BlockTag {
+    /// The latest mined block.
+    #[default]
+    #[serde(rename = "latest")]
+    Latest,
+    /// The latest safe head block.
+    /// See
+    /// <https://www.alchemy.com/overviews/ethereum-commitment-levels#what-are-ethereum-commitment-levels>
+    #[serde(rename = "safe")]
+    Safe,
+    /// The latest finalized block.
+    /// See
+    /// <https://www.alchemy.com/overviews/ethereum-commitment-levels#what-are-ethereum-commitment-levels>
+    #[serde(rename = "finalized")]
+    Finalized,
+    #[serde(rename = "earliest")]
+    Earliest,
+    #[serde(rename = "pending")]
+    Pending,
+}
+
+impl Display for BlockTag {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Latest => write!(f, "latest"),
+            Self::Safe => write!(f, "safe"),
+            Self::Finalized => write!(f, "finalized"),
+            Self::Earliest => write!(f, "earliest"),
+            Self::Pending => write!(f, "pending"),
+        }
+    }
+}
+
+/// The block specification indicating which block to query.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(untagged)]
+pub enum BlockSpec {
+    /// Query the block with the specified index.
+    Number(BlockNumber),
+    /// Query the block with the specified tag.
+    Tag(BlockTag),
+}
+
+impl Default for BlockSpec {
+    fn default() -> Self {
+        Self::Tag(BlockTag::default())
+    }
+}
+
+impl std::str::FromStr for BlockSpec {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.starts_with("0x") {
+            let block_number = BlockNumber::from_str_hex(s)
+                .map_err(|e| format!("failed to parse block number '{s}': {e}"))?;
+            return Ok(BlockSpec::Number(block_number));
+        }
+        Ok(BlockSpec::Tag(match s {
+            "latest" => BlockTag::Latest,
+            "safe" => BlockTag::Safe,
+            "finalized" => BlockTag::Finalized,
+            _ => return Err(format!("unknown block tag '{s}'")),
```

Comment:
Are "earliest" and "pending" considered "unknown block tags"?

Reply:
removed [here](https://github.com/internet-computer-protocol/evm-rpc-canister/pull/279/commits/06a0614692f74ebc0ab81040401b30e336de53c8)